libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -2980,6 +2953,80 @@ string YulUtilFunctions::conversionFunction(Type const& _from, Type const& _to)
 	});
 }
 
+string YulUtilFunctions::arrayConversionFunction(ArrayType const& _from, ArrayType const& _to)
+{
+	string functionName =
+		"convert_array_" +
+		_from.identifier() +
+		"_to_" +
+		_to.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		Whiskers templ(R"(
+			function <functionName>(value<?fromCalldataDynamic>, length</fromCalldataDynamic>) -> converted {
+				<?sameTypes>
+					converted := value
+				<!sameTypes>
+					<?toMemory>
+						// Copy the array to a free position in memory, unless it is already in memory.
+						<?fromMemory>
+							converted := value
+						</fromMemory>
+						<?fromStorage>
+							converted := <arrayStorageToMem>(value)
+						</fromStorage>
+						<?fromCalldata>
+							converted := <allocateMemoryArray>(length)
+							<copyToMemory>(value, add(converted, 0x20), length)
```

Comment:
Is it possible that `fromCalldata` is true but `fromCalldataDynamic` is false? In that case `length` would be undeclared

Reply:
Yes, when array is static. Now it is not an issue as there is assertion that this copy is supported only for dynamic arrays. Next PR (#10211) solves this. If you want, I can add changes from that PR to this one so we can solve all issues at once?

Reply:
I think it's fine to keep the PRs separate. But is there an assertion in this PR preventing that case?

Reply:
Ah, actually there is not, I checked now more carefully. Added new assertion to make sure now.